lucene/core/src/java/org/apache/lucene/codecs/lucene90/PForUtil.java

```diff
@@ -121,4 +167,146 @@ void skip(DataInput in) throws IOException {
       in.skipBytes(forUtil.numBytes(bitsPerValue) + (numExceptions << 1));
     }
   }
+
+  /**
+   * Fill {@code longs} with the final values for the case of all deltas being 1. Note this assumes
+   * there are no exceptions to apply.
+   */
+  private static void prefixSumOfOnes(long[] longs, long base) {
+    System.arraycopy(IDENTITY_PLUS_ONE, 0, longs, 0, ForUtil.BLOCK_SIZE);
+    // This loop gets auto-vectorized
+    for (int i = 0; i < ForUtil.BLOCK_SIZE; ++i) {
+      longs[i] += base;
+    }
+  }
+
+  /**
+   * Fill {@code longs} with the final values for the case of all deltas being {@code val}. Note
+   * this assumes there are no exceptions to apply.
+   */
+  private static void prefixSumOf(long[] longs, long base, long val) {
+    for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
+      longs[i] = (i + 1) * val + base;
```

Comment:
have you looked into whether the JVM auto-vectorized this? If it doesn't I wonder if we could trick it into vectorizing by doing something like that instead

```suggestion
      longs[i] = IDENTITY_PLUS_ONE[i] * val + base;
```

Reply:
I don't actually know how to check this. I tried looking at the bytecode, but that's not what we want right? We need to see the actual compiled code to see if it's making use of vectorized instructions? Do you have any tips on how to look at this? I'd love to learn!

Reply:
@gsmiller here's what i do on linux:

you need to build `hsdis-amd64.so` or similar (it is in openjdk sources and instructions to build from source can be found online, maybe even packages for your OS) and drop it into jvm's `lib/server` directory. This is a hook to binutils and allows for low level PrintAssembly option to give useful output.

it may be easiest to write jmh microbenchmark targeting the code in question, and use jmh `perfasm`, that's what i do. it prevents mistakes by ensuring code is really "hot", and also targets only the interesting parts so you don't have to sort through tons of assembly. also on the side you get a benchmark for making tweaks and confirming them :)

Reply:
here's my notes for building the shared library from last time i recently did this (may be incomplete, sorry):

```
$ git clone git@github.com:openjdk/jdk.git
$ cd jdk/src/utils/hsdis/build
$ curl -fLO https://ftp.gnu.org/gnu/binutils/binutils-2.36.tar.gz
$ gzip -dc binutils-2.36.tar.gz | tar -xvf -
$ mv binutils-2.36 binutils
$ cd ..
$ make ARCH=amd64
```

This will create build/linux-amd64/hsdis-amd64.so
For me it initially fails because of binutils enabling `-Werror` (i have a newer gcc), so i had to hack the Makefile for it to really work:
```
--- a/src/utils/hsdis/Makefile
+++ b/src/utils/hsdis/Makefile
@@ -192,7 +192,7 @@ $(LIBRARIES): $(TARGET_DIR) $(TARGET_DIR)/Makefile
        if [ ! -f $@ ]; then cd $(TARGET_DIR); make all-opcodes; fi

 $(TARGET_DIR)/Makefile:
-       (cd $(TARGET_DIR); CC=$(CC) CFLAGS="$(CFLAGS)" AR="$(AR)" $(BINUTILSDIR)/configure --disable-nls $(CONFIGURE_ARGS))
+       (cd $(TARGET_DIR); CC=$(CC) CFLAGS="$(CFLAGS)" AR="$(AR)" $(BINUTILSDIR)/configure --disable-werror --disable-nls $(CONFIGURE_ARGS))
```

Now you can print assembly generated from the JIT, but you need to ensure the code is really fully C2-compiled so that you are looking at the correct stuff, etc.... essentially target the code in question and heat it up. So I use JMH for that, they have instructions in the repo: https://github.com/openjdk/jmh

Reply:
Awesome, thanks @rmuir! I'll dig into these suggestions and see what I can turn up. Cheers!

Reply:
Well, I've made less progress than I would have liked in being able to answer whether-or-not this loop is getting auto-vectorized. To be honest, I've been stumbling along just trying to get to the assembly. I _think_ I've got the assembly for both the `prefixSumOf` and `prefixSumOfOnes` (as a comparison), but this is from my 2019 MBP, not a "proper" linux box (which seems to have bitten the dust overnight... so sorting that out). But now I'm not really sure what I'm looking for. I'm going to keep trying to sort this out on my own, but since it's a little slow-going, I'm pasting the results here in case someone more well-versed in this wants to have a look in the meantime.

I grabbed the following using `java -XX:+UnlockDiagnosticVMOptions -XX:CompileCommand="print jpountz.PForDeltaDecoder::prefixSum*" -jar target/benchmarks.jar` on a [microbenchmark](https://github.com/gsmiller/decode-128-ints-benchmark/tree/pfor-delta) I setup from forking earlier work by @jpountz. So I think the code should be "hot" here (and for whatever reason, I've been having trouble getting `-prof perfasm` to work on the bench run, which is another thing I'm trying to sort through):

prefixSumOf (the method in question):
```
----------------------------------------------------------------------
jpountz/PForDeltaDecoder.prefixSumOf(J[JJ)V  [0x0000000129eaed20, 0x0000000129eaee38]  280 bytes
[Entry Point]
[Verified Entry Point]
[Constants]
  # {method} {0x0000000135b8b150} 'prefixSumOf' '(J[JJ)V' in 'jpountz/PForDeltaDecoder'
  # parm0:    rsi:rsi   = long
  # parm1:    rdx:rdx   = '[J'
  # parm2:    rcx:rcx   = long
  #           [sp+0x30]  (sp of caller)
  0x0000000129eaed20: mov    %eax,-0x14000(%rsp)
  0x0000000129eaed27: push   %rbp
  0x0000000129eaed28: sub    $0x20,%rsp         ;*synchronization entry
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@-1 (line 28)

  0x0000000129eaed2c: mov    0xc(%rdx),%r10d    ; implicit exception: dispatches to 0x0000000129eaedf6
  0x0000000129eaed30: test   %r10d,%r10d
  0x0000000129eaed33: jbe    0x0000000129eaedf6
  0x0000000129eaed39: cmp    $0x7f,%r10d
  0x0000000129eaed3d: jbe    0x0000000129eaedf6
  0x0000000129eaed43: mov    %rsi,%r10
  0x0000000129eaed46: add    %rcx,%r10
  0x0000000129eaed49: mov    %r10,0x10(%rdx)    ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@23 (line 29)

  0x0000000129eaed4d: mov    $0x1,%r11d
  0x0000000129eaed53: nopw   0x0(%rax,%rax,1)
  0x0000000129eaed5c: data32 data32 xchg %ax,%ax  ;*aload_2 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@11 (line 29)

  0x0000000129eaed60: movslq %r11d,%r10
  0x0000000129eaed63: mov    %r10,%r8
  0x0000000129eaed66: add    $0x1,%r8
  0x0000000129eaed6a: imul   %rsi,%r8
  0x0000000129eaed6e: add    %rcx,%r8
  0x0000000129eaed71: mov    %r8,0x10(%rdx,%r10,8)
  0x0000000129eaed76: mov    %r10,%r8
  0x0000000129eaed79: add    $0x4,%r8
  0x0000000129eaed7d: imul   %rsi,%r8
  0x0000000129eaed81: add    %rcx,%r8
  0x0000000129eaed84: mov    %r10,%r9
  0x0000000129eaed87: add    $0x3,%r9
  0x0000000129eaed8b: imul   %rsi,%r9
  0x0000000129eaed8f: add    %rcx,%r9
  0x0000000129eaed92: mov    %r10,%rbx
  0x0000000129eaed95: add    $0x2,%rbx
  0x0000000129eaed99: imul   %rsi,%rbx
  0x0000000129eaed9d: add    %rcx,%rbx
  0x0000000129eaeda0: mov    %rbx,0x18(%rdx,%r10,8)
  0x0000000129eaeda5: mov    %r9,0x20(%rdx,%r10,8)
  0x0000000129eaedaa: mov    %r8,0x28(%rdx,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@23 (line 29)

  0x0000000129eaedaf: add    $0x4,%r11d         ;*iadd {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@17 (line 29)

  0x0000000129eaedb3: cmp    $0x7d,%r11d
  0x0000000129eaedb7: jl     0x0000000129eaed60  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@8 (line 28)

  0x0000000129eaedb9: cmp    $0x80,%r11d
  0x0000000129eaedc0: jge    0x0000000129eaede6
  0x0000000129eaedc2: xchg   %ax,%ax            ;*aload_2 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@11 (line 29)

  0x0000000129eaedc4: movslq %r11d,%r10
  0x0000000129eaedc7: mov    %r10,%r8
  0x0000000129eaedca: add    $0x1,%r8
  0x0000000129eaedce: imul   %rsi,%r8
  0x0000000129eaedd2: add    %rcx,%r8
  0x0000000129eaedd5: mov    %r8,0x10(%rdx,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@23 (line 29)

  0x0000000129eaedda: inc    %r11d              ;*iadd {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@17 (line 29)

  0x0000000129eaeddd: cmp    $0x80,%r11d
  0x0000000129eaede4: jl     0x0000000129eaedc4
  0x0000000129eaede6: add    $0x20,%rsp
  0x0000000129eaedea: pop    %rbp
  0x0000000129eaedeb: mov    0x108(%r15),%r10
  0x0000000129eaedf2: test   %eax,(%r10)        ;   {poll_return}
  0x0000000129eaedf5: retq                      ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@8 (line 28)

  0x0000000129eaedf6: mov    %rsi,%rbp
  0x0000000129eaedf9: mov    %rdx,(%rsp)
  0x0000000129eaedfd: mov    %rcx,0x8(%rsp)
  0x0000000129eaee02: mov    $0xffffff7e,%esi
  0x0000000129eaee07: callq  0x00000001223e1b00  ; ImmutableOopMap{[0]=Oop }
                                                ;*if_icmpge {reexecute=1 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@8 (line 28)
                                                ;   {runtime_call UncommonTrapBlob}
  0x0000000129eaee0c: hlt    
  0x0000000129eaee0d: hlt    
  0x0000000129eaee0e: hlt    
  0x0000000129eaee0f: hlt    
  0x0000000129eaee10: hlt    
  0x0000000129eaee11: hlt    
  0x0000000129eaee12: hlt    
  0x0000000129eaee13: hlt    
  0x0000000129eaee14: hlt    
  0x0000000129eaee15: hlt    
  0x0000000129eaee16: hlt    
  0x0000000129eaee17: hlt    
  0x0000000129eaee18: hlt    
  0x0000000129eaee19: hlt    
  0x0000000129eaee1a: hlt    
  0x0000000129eaee1b: hlt    
  0x0000000129eaee1c: hlt    
  0x0000000129eaee1d: hlt    
  0x0000000129eaee1e: hlt    
  0x0000000129eaee1f: hlt    
[Exception Handler]
[Stub Code]
  0x0000000129eaee20: jmpq   0x0000000122417200  ;   {no_reloc}
[Deopt Handler Code]
  0x0000000129eaee25: callq  0x0000000129eaee2a
  0x0000000129eaee2a: subq   $0x5,(%rsp)
  0x0000000129eaee2f: jmpq   0x00000001223e1720  ;   {runtime_call DeoptimizationBlob}
  0x0000000129eaee34: hlt    
  0x0000000129eaee35: hlt    
  0x0000000129eaee36: hlt    
  0x0000000129eaee37: hlt  
```

prefixSumOfOnes (for comparison):
```
----------------------------------------------------------------------
jpountz/PForDeltaDecoder.prefixSumOfOnes([JJ)V  [0x00000001151e2320, 0x00000001151e24f8]  472 bytes
[Entry Point]
[Verified Entry Point]
[Constants]
  # {method} {0x0000000126ebb048} 'prefixSumOfOnes' '([JJ)V' in 'jpountz/PForDeltaDecoder'
  # parm0:    rsi:rsi   = '[J'
  # parm1:    rdx:rdx   = long
  #           [sp+0x30]  (sp of caller)
  0x00000001151e2320: mov    %eax,-0x14000(%rsp)
  0x00000001151e2327: push   %rbp
  0x00000001151e2328: sub    $0x20,%rsp         ;*synchronization entry
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@-1 (line 20)

  0x00000001151e232c: mov    %rdx,%rbx
  0x00000001151e232f: mov    %rsi,%r13
  0x00000001151e2332: mov    0xc(%rsi),%ebp     ; implicit exception: dispatches to 0x00000001151e24cc
  0x00000001151e2335: cmp    $0x80,%ebp
  0x00000001151e233b: jb     0x00000001151e24a4
  0x00000001151e2341: add    $0x10,%rsi
  0x00000001151e2345: mov    $0x80,%edx
  0x00000001151e234a: movabs $0x70f829c00,%rdi  ;   {oop([J{0x000000070f829c00})}
  0x00000001151e2354: add    $0x10,%rdi
  0x00000001151e2358: vzeroupper 
  0x00000001151e235b: movabs $0x10d721460,%r10
  0x00000001151e2365: callq  *%r10              ;*invokestatic arraycopy {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@9 (line 20)

  0x00000001151e2368: test   %ebp,%ebp
  0x00000001151e236a: jbe    0x00000001151e24b8
  0x00000001151e2370: cmp    $0x7f,%ebp
  0x00000001151e2373: jbe    0x00000001151e24b8
  0x00000001151e2379: mov    %r13d,%r11d
  0x00000001151e237c: shr    $0x3,%r11d
  0x00000001151e2380: and    $0x3,%r11d
  0x00000001151e2384: xor    %r10d,%r10d
  0x00000001151e2387: mov    $0x1,%r8d
  0x00000001151e238d: sub    %r11d,%r8d
  0x00000001151e2390: and    $0x3,%r8d
  0x00000001151e2394: inc    %r8d
  0x00000001151e2397: mov    $0x80,%r11d
  0x00000001151e239d: cmp    %r11d,%r8d
  0x00000001151e23a0: cmovg  %r11d,%r8d         ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@21 (line 23)

  0x00000001151e23a4: add    %rbx,0x10(%r13,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@27 (line 23)

  0x00000001151e23a9: inc    %r10d              ;*iinc {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@28 (line 22)

  0x00000001151e23ac: cmp    %r8d,%r10d
  0x00000001151e23af: jl     0x00000001151e23a4  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@18 (line 22)

  0x00000001151e23b1: vmovq  %rbx,%xmm0
  0x00000001151e23b6: vpunpcklqdq %xmm0,%xmm0,%xmm0
  0x00000001151e23ba: (bad)  
  0x00000001151e23be: rolb   $0xc4,(%rcx)       ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@21 (line 23)

  0x00000001151e23c1: cmpl   $0xc410d54c,-0x2c(%rbp)
  0x00000001151e23c8: cmpl   $0xc410d54c,0x7f(%rsi)
  0x00000001151e23cf: cmpl   $0xc430d54c,-0x2c(%rbp)
  0x00000001151e23d6: cmpl   $0xc430d54c,0x7f(%rsi)
  0x00000001151e23dd: cmpl   $0xc450d54c,-0x2c(%rbp)
  0x00000001151e23e4: cmpl   $0xc450d54c,0x7f(%rsi)
  0x00000001151e23eb: cmpl   $0xc470d54c,-0x2c(%rbp)
  0x00000001151e23f2: cmpl   $0xc470d54c,0x7f(%rsi)
  0x00000001151e23f9: cmpl   $0x90d58c,-0x2c(%rbp)
  0x00000001151e2400: add    %al,(%rax)
  0x00000001151e2402: vmovdqu %ymm1,0x90(%r13,%r10,8)
  0x00000001151e240c: (bad)  
  0x00000001151e2410: mov    %ss,%ebp
  0x00000001151e2412: mov    $0x0,%al
  0x00000001151e2414: add    %al,(%rax)
  0x00000001151e2416: vmovdqu %ymm1,0xb0(%r13,%r10,8)
  0x00000001151e2420: (bad)  
  0x00000001151e2424: mov    %ss,%ebp
  0x00000001151e2426: rolb   (%rax)
  0x00000001151e2428: add    %al,(%rax)
  0x00000001151e242a: vmovdqu %ymm1,0xd0(%r13,%r10,8)
  0x00000001151e2434: (bad)  
  0x00000001151e2438: mov    %ss,%ebp
  0x00000001151e243a: lock add %al,(%rax)
  0x00000001151e243d: add    %al,%ah
  0x00000001151e243f: cmpl   $0xf0d58c,0x7f(%rsi)
  0x00000001151e2446: add    %al,(%rax)         ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@27 (line 23)

  0x00000001151e2448: add    $0x20,%r10d        ;*iinc {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@28 (line 22)

  0x00000001151e244c: cmp    $0x61,%r10d
  0x00000001151e2450: jl     0x00000001151e23c0  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@18 (line 22)

  0x00000001151e2456: cmp    $0x7d,%r10d
  0x00000001151e245a: jge    0x00000001151e2474  ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@21 (line 23)

  0x00000001151e245c: (bad)  
  0x00000001151e2460: rex.WR (bad) 
  0x00000001151e2462: adc    %al,%ah
  0x00000001151e2464: cmpl   $0x4110d54c,0x7f(%rsi)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@27 (line 23)

  0x00000001151e246b: add    $0x4,%edx          ;*iinc {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@28 (line 22)

  0x00000001151e246e: cmp    $0x7d,%r10d
  0x00000001151e2472: jl     0x00000001151e245c  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@18 (line 22)

  0x00000001151e2474: cmp    $0x80,%r10d
  0x00000001151e247b: jge    0x00000001151e2491
  0x00000001151e247d: data32 xchg %ax,%ax       ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@21 (line 23)

  0x00000001151e2480: add    %rbx,0x10(%r13,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@27 (line 23)

  0x00000001151e2485: inc    %r10d              ;*iinc {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@28 (line 22)

  0x00000001151e2488: cmp    $0x80,%r10d
  0x00000001151e248f: jl     0x00000001151e2480  ;*synchronization entry
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@-1 (line 20)

  0x00000001151e2491: vzeroupper 
  0x00000001151e2494: add    $0x20,%rsp
  0x00000001151e2498: pop    %rbp
  0x00000001151e2499: mov    0x108(%r15),%r10
  0x00000001151e24a0: test   %eax,(%r10)        ;   {poll_return}
  0x00000001151e24a3: retq   
  0x00000001151e24a4: mov    $0xffffffcc,%esi
  0x00000001151e24a9: mov    %r13,%rbp
  0x00000001151e24ac: mov    %rdx,(%rsp)
  0x00000001151e24b0: vzeroupper 
  0x00000001151e24b3: callq  0x000000010d713b00  ; ImmutableOopMap{rbp=Oop }
                                                ;*invokestatic arraycopy {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@9 (line 20)
                                                ;   {runtime_call UncommonTrapBlob}
  0x00000001151e24b8: mov    $0xffffff7e,%esi
  0x00000001151e24bd: mov    %r13,%rbp
  0x00000001151e24c0: mov    %rbx,(%rsp)
  0x00000001151e24c4: vzeroupper 
  0x00000001151e24c7: callq  0x000000010d713b00  ; ImmutableOopMap{rbp=Oop }
                                                ;*if_icmpge {reexecute=1 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@18 (line 22)
                                                ;   {runtime_call UncommonTrapBlob}
  0x00000001151e24cc: mov    $0xfffffff6,%esi
  0x00000001151e24d1: data32 xchg %ax,%ax
  0x00000001151e24d4: vzeroupper 
  0x00000001151e24d7: callq  0x000000010d713b00  ; ImmutableOopMap{}
                                                ;*invokestatic arraycopy {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@9 (line 20)
                                                ;   {runtime_call UncommonTrapBlob}
  0x00000001151e24dc: hlt    
  0x00000001151e24dd: hlt    
  0x00000001151e24de: hlt    
  0x00000001151e24df: hlt    
[Exception Handler]
[Stub Code]
  0x00000001151e24e0: jmpq   0x000000010d749200  ;   {no_reloc}
[Deopt Handler Code]
  0x00000001151e24e5: callq  0x00000001151e24ea
  0x00000001151e24ea: subq   $0x5,(%rsp)
  0x00000001151e24ef: jmpq   0x000000010d713720  ;   {runtime_call DeoptimizationBlob}
  0x00000001151e24f4: hlt    
  0x00000001151e24f5: hlt    
  0x00000001151e24f6: hlt    
  0x00000001151e24f7: hlt  
```

Reply:
Oh, for what it's worth, I pulled in the suggestion anyway to use `IDENTITY_PLUS_ONE[i]` in place of `(i + 1)`. No matter what, I think it makes sense to access this static array vs. computing `i + 1` in each loop...

Reply:
> I've been having trouble getting `-prof perfasm` to work on the bench run, which is another thing I'm trying to sort through)

This will only work on linux, since it uses Linux perf.

Reply:
1. `yum install perf`
2. stop banging head against wall
:)

Reply:
Hmm... I'm not getting anything useful out of `perfasm`. I've got `perf` installed and running on linux, but no matter what I seem to do, I just get the following (I've tried lots of different settings for `hotThreshold` and `frequency`, and have also increased the number of benchmark iterations and time-per-iteration in various ways):
```
Secondary result "jpountz.PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder:·asm":
PrintAssembly processed: 361476 total address lines.
Perf output processed (skipped 16.036 seconds):
 Column 1: cycles (0 events)

WARNING: No hottest code region above the threshold (10.00%) for disassembly.
Use "hotThreshold" profiler option to lower the filter threshold.

....[Hottest Regions]...............................................................................
....................................................................................................
         <totals>

....[Hottest Methods (after inlining)]..............................................................
....................................................................................................
         <totals>

....[Distribution by Source]........................................................................
....................................................................................................
         <totals>

WARNING: The perf event count is suspiciously low (0). The performance data might be
inaccurate or misleading. Try to do the profiling again, or tune up the sampling frequency.
With some profilers on Mac OS X, System Integrity Protection (SIP) may prevent profiling.
In such case, temporarily disabling SIP with 'csrutil disable' might help.
```

I'm wondering if this is due to running on a virtual machine (AWS ec2 host)? I came across this [read](http://psy-lob-saw.blogspot.com/2015/07/jmh-perfasm.html) that seemed to indicate the need to run on "real hardware." I tried @jpountz original benchmark branch as well with `-prof perfasm` and get the same issue. I've also confirmed that `-XX:+PrintAssembly` is working on this machine (with `hsdis-amd64.so` under `JAVA_HOME/lib/server`). I'm calling it for today but will see if I can make more progress tomorrow. If I'm doing something obviously wrong here, please point it out! Thanks again for the help!

Reply:
@gsmiller my first guess would be permissions related. I'd first try `sudo perf top` then `perf top` and see if one or both don't work. If that's the case, i'd first inspect sysctl to see: `sysctl kernel.perf_event_paranoid`. I set `kernel.perf_event_paranoid = -1` because I want to profile things :)

Reply:
Hmm, thanks @rmuir. That was indeed one problem at least. I confirmed that `perf top` was giving me permission errors (while `sudo perf top` was not). I updated `kernel.perf_event_paranoid = -1` and confirmed `perf top` now works, so that's great! But I'm still seeing the same problem when trying to get `perfasm` working. So progress, but still not there yet unfortunately. Thanks again for the pointers!

Reply:
> I came across this [read](http://psy-lob-saw.blogspot.com/2015/07/jmh-perfasm.html) that seemed to indicate the need to run on "real hardware."

I dont know about this. For me it works fine both on real hardware and inside QEMU vm, i see similar results for a simple test command passing the same arguments that jmh does behind the scenes:
```
perf record --freq 1000 --event cycles,instructions --output test.tmp dd if=/dev/urandom of=/dev/null bs=1M count=100
```
If it says 0 samples, then something is off.

hardware wise, I see the following config:
```
host$ journalctl -k | grep PMU
kernel: Performance Events: PEBS fmt3+, Skylake events, 32-deep LBR, full-width counters, Intel PMU driver.

guest$ journalctl -k | grep PMU
kernel: Performance Events: Skylake events, full-width counters, Intel PMU driver.
```

Reply:
just looking into some blogs, you might need to switch instance type (maybe dedicated instance) to have a vPMU on ec2: https://oavdeev.github.io/posts/vpmu_support_z1d/

Reply:
Thanks, I'll have a look at that link and see about setting up a different instance type.

In the meantime, for what it's worth, I used `PrintAssembly` and was able to get the assembly for both `prefixSumOf` and `prefixSumOfOnes` on my linux box. Same thing I did on my MBP, but on the linux machine. In `prefixSumOfOnes`, I'm seeing this block, which looks an awful lot like vectorized instructions:
```
  0x00007f62379d6831: vmovq  %rbx,%xmm0
  0x00007f62379d6836: vpunpcklqdq %xmm0,%xmm0,%xmm0
  0x00007f62379d683a: vinserti128 $0x1,%xmm0,%ymm0,%ymm0
                                                ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@21 (line 23)

  0x00007f62379d6840: vpaddq 0x10(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d6847: vmovdqu %ymm1,0x10(%r13,%r10,8)
  0x00007f62379d684e: vpaddq 0x30(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d6855: vmovdqu %ymm1,0x30(%r13,%r10,8)
  0x00007f62379d685c: vpaddq 0x50(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d6863: vmovdqu %ymm1,0x50(%r13,%r10,8)
  0x00007f62379d686a: vpaddq 0x70(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d6871: vmovdqu %ymm1,0x70(%r13,%r10,8)
  0x00007f62379d6878: vpaddq 0x90(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d6882: vmovdqu %ymm1,0x90(%r13,%r10,8)
  0x00007f62379d688c: vpaddq 0xb0(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d6896: vmovdqu %ymm1,0xb0(%r13,%r10,8)
  0x00007f62379d68a0: vpaddq 0xd0(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d68aa: vmovdqu %ymm1,0xd0(%r13,%r10,8)
  0x00007f62379d68b4: vpaddq 0xf0(%r13,%r10,8),%ymm0,%ymm1
  0x00007f62379d68be: vmovdqu %ymm1,0xf0(%r13,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
                                                ; - jpountz.PForDeltaDecoder::prefixSumOfOnes@27 (line 23)
```

Unfortunately, I'm not seeing the same thing in `prefixSumOf` (for the following implementation):
```
    private static void prefixSumOf(long val, long[] arr, long base) {
        for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
            arr[i] = IDENTITY_PLUS_ONE[i] * val + base;
        }
    }
```

So in trying to answer the original question here of whether-or-not this code is getting auto-vectorized, I'm guessing it's not based on what I'm seeing, but I'd still like to get `perfasm` working. It's entirely possible I'm misinterpreting the assembly as well. I can post the full decompiled assembly of both methods here if interested. It sure will be nice when we can just ask Java to do this explicitly...

Finally, I also tried rewriting this to:
```
    private static void prefixSumOf(long val, long[] arr, long base) {
        System.arraycopy(IDENTITY_PLUS_ONE, 0, arr, 0, ForUtil.BLOCK_SIZE);
        for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
            arr[i] *= val;
        }
        for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
            arr[i] += base;
        }
    }
```
With this implementation, it looks like the second loop is getting auto-vectorized but it doesn't appear to me like the first loop is. I ran this "two loop" approach against the original "single loop" implementation in a microbench, and it seemed to perform slightly worse (but not by a lot).

Reply:
> I'm seeing this block, which looks an awful lot like vectorized instructions:

yes, the first part looks like ReplL_reg "Replicate long (8 byte) scalar to be vector" from hotspot with avx256 (see the x86.ad file in jdk src).

Reply:
@rmuir instance type swap did the trick. Thanks so much! Grabbing your command from above, here's the suspicious difference (in case this helps in the future):
* instance type: m4.10xlarge
```
m4.10xl% sudo journalctl -k | grep PMU                                                                                                                                                                    
Mar 24 15:02:23 localhost kernel: Performance Events: unsupported p6 CPU model 63 no PMU driver, software events only.
Mar 24 15:02:23 localhost kernel: RAPL PMU: API unit is 2^-32 Joules, 0 fixed counters, 655360 ms ovfl timer
```
* instance type: m5.12xlarge
```
m5.12xl% sudo journalctl -k | grep PMU                                                                                                                
Apr 08 04:02:43 localhost kernel: Performance Events: Skylake events, Intel PMU driver.
Apr 08 04:02:43 localhost kernel: RAPL PMU: API unit is 2^-32 Joules, 0 fixed counters, 10737418240 ms ovfl timer
```
I've confirmed `perfasm` is working for me. I'll get some results updated here shortly. Thanks again!

Reply:
@gsmiller glad to hear you are up and running! We need more eyes on this stuff, and they don't exactly make it easy!

Reply:
Here's what I've found with `perfasm` on this microbenchmark branch:
1. The `prefixSumOf` method in question [1] is _not_ auto-vectorizing. The assembly loop is below [2].
2. If I change the implementation of `prefixSumOf` to use two loops [3], the second "add" loop is auto-vectoring in the same way that `prefixSumOfOnes` does [4], but the first "multiply" loop does not [5].
3. Even though the second approach [3] gets partially vectorized, it's significantly less performant than the vanilla, single-loop approach [6].

[1]
```
    private static void prefixSumOf(long val, long[] arr, long base) {
        for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
            arr[i] = IDENTITY_PLUS_ONE[i] * val + base;
        }
    }
```
[2]
```
  0.45%    ↗  0x00007f37dfa02c52: mov    %r9,%r8
  0.30%    │  0x00007f37dfa02c55: movabs $0xd2816340,%rdi   ;   {oop([J{0x00000000d2816340})}
  3.16%    │  0x00007f37dfa02c5f: imul   0x10(%rdi,%r10,8),%r8
  1.12%    │  0x00007f37dfa02c65: add    %rcx,%r8
  1.42%    │  0x00007f37dfa02c68: mov    %r8,0x10(%rbx,%r10,8)
  2.97%    │  0x00007f37dfa02c6d: mov    %r9,%r8
  2.62%    │  0x00007f37dfa02c70: imul   0x18(%rdi,%r10,8),%r8
  1.37%    │  0x00007f37dfa02c76: add    %rcx,%r8
  1.40%    │  0x00007f37dfa02c79: mov    %r8,0x18(%rbx,%r10,8)
  5.71%    │  0x00007f37dfa02c7e: mov    %r9,%r8
  2.02%    │  0x00007f37dfa02c81: imul   0x20(%rdi,%r10,8),%r8
  1.08%    │  0x00007f37dfa02c87: add    %rcx,%r8
  1.91%    │  0x00007f37dfa02c8a: mov    %r8,0x20(%rbx,%r10,8)
  4.96%    │  0x00007f37dfa02c8f: mov    %r9,%r8
  1.57%    │  0x00007f37dfa02c92: imul   0x28(%rdi,%r10,8),%r8
  0.71%    │  0x00007f37dfa02c98: add    %rcx,%r8
  0.56%    │  0x00007f37dfa02c9b: mov    %r8,0x28(%rbx,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
           │                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@24 (line 29)
           │                                                ; - jpountz.PForDeltaDecoder::decodeAndPrefixSum@32 (line 59)
           │                                                ; - jpountz.PackedIntsDeltaDecodeBenchmark::pForDeltaDecoder@42 (line 29)
           │                                                ; - jpountz.generated.PackedIntsDeltaDecodeBenchmark_pForDeltaDecoder_jmhTest::pForDeltaDecoder_thrpt_jmhStub@151 (line 240)
  4.79%    │  0x00007f37dfa02ca0: add    $0x4,%r10d         ;*iinc {reexecute=0 rethrow=0 return_oop=0}
           │                                                ; - jpountz.PForDeltaDecoder::prefixSumOf@25 (line 28)
           │                                                ; - jpountz.PForDeltaDecoder::decodeAndPrefixSum@32 (line 59)
           │                                                ; - jpountz.PackedIntsDeltaDecodeBenchmark::pForDeltaDecoder@42 (line 29)
           │                                                ; - jpountz.generated.PackedIntsDeltaDecodeBenchmark_pForDeltaDecoder_jmhTest::pForDeltaDecoder_thrpt_jmhStub@151 (line 240)
  1.22%    │  0x00007f37dfa02ca4: cmp    $0x7d,%r10d
           ╰  0x00007f37dfa02ca8: jl     0x00007f37dfa02c52  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
```
[3]
```
    private static void prefixSumOfTwoLoops(long val, long[] arr, long base) {
        System.arraycopy(IDENTITY_PLUS_ONE, 0, arr, 0, ForUtil.BLOCK_SIZE);
        for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
            arr[i] *= val;
        }
        for (int i = 0; i < ForUtil.BLOCK_SIZE; i++) {
            arr[i] += base;
        }
    }
```
[4]
```
  0.11%       ↗    0x00007f1607a05810: vpaddq 0x10(%rbp,%r11,8),%ymm0,%ymm1
  0.17%       │    0x00007f1607a05817: vmovdqu %ymm1,0x10(%rbp,%r11,8)
  0.24%       │    0x00007f1607a0581e: vpaddq 0x30(%rbp,%r11,8),%ymm0,%ymm1
  0.20%       │    0x00007f1607a05825: vmovdqu %ymm1,0x30(%rbp,%r11,8)
  0.15%       │    0x00007f1607a0582c: vpaddq 0x50(%rbp,%r11,8),%ymm0,%ymm1
  0.30%       │    0x00007f1607a05833: vmovdqu %ymm1,0x50(%rbp,%r11,8)
  0.09%       │    0x00007f1607a0583a: vpaddq 0x70(%rbp,%r11,8),%ymm0,%ymm1
  0.30%       │    0x00007f1607a05841: vmovdqu %ymm1,0x70(%rbp,%r11,8)
              │    0x00007f1607a05848: vpaddq 0x90(%rbp,%r11,8),%ymm0,%ymm1
  0.15%       │    0x00007f1607a05852: vmovdqu %ymm1,0x90(%rbp,%r11,8)
  0.20%       │    0x00007f1607a0585c: vpaddq 0xb0(%rbp,%r11,8),%ymm0,%ymm1
  0.11%       │    0x00007f1607a05866: vmovdqu %ymm1,0xb0(%rbp,%r11,8)
  0.09%       │    0x00007f1607a05870: vpaddq 0xd0(%rbp,%r11,8),%ymm0,%ymm1
  0.37%       │    0x00007f1607a0587a: vmovdqu %ymm1,0xd0(%rbp,%r11,8)
  0.07%       │    0x00007f1607a05884: vpaddq 0xf0(%rbp,%r11,8),%ymm0,%ymm1
  0.26%       │    0x00007f1607a0588e: vmovdqu %ymm1,0xf0(%rbp,%r11,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
              │                                                  ; - jpountz.PForDeltaDecoder::prefixSumOfTwoLoops@55 (line 39)
              │                                                  ; - jpountz.PForDeltaDecoder::decodeAndPrefixSum@32 (line 59)
              │                                                  ; - jpountz.PackedIntsDeltaDecodeBenchmark::pForDeltaDecoder@42 (line 29)
              │                                                  ; - jpountz.generated.PackedIntsDeltaDecodeBenchmark_pForDeltaDecoder_jmhTest::pForDeltaDecoder_thrpt_jmhStub@151 (line 240)
  0.15%       │    0x00007f1607a05898: add    $0x20,%r11d        ;*iinc {reexecute=0 rethrow=0 return_oop=0}
              │                                                  ; - jpountz.PForDeltaDecoder::prefixSumOfTwoLoops@56 (line 38)
              │                                                  ; - jpountz.PForDeltaDecoder::decodeAndPrefixSum@32 (line 59)
              │                                                  ; - jpountz.PackedIntsDeltaDecodeBenchmark::pForDeltaDecoder@42 (line 29)
              │                                                  ; - jpountz.generated.PackedIntsDeltaDecodeBenchmark_pForDeltaDecoder_jmhTest::pForDeltaDecoder_thrpt_jmhStub@151 (line 240)
  0.13%       │    0x00007f1607a0589c: cmp    $0x61,%r11d
              ╰    0x00007f1607a058a0: jl     0x00007f1607a05810  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
```
[5]
```
  0.31%   ↗        0x00007f1607a05750: mov    %r13,%r11
  1.06%   │        0x00007f1607a05753: imul   0x10(%rbp,%r10,8),%r11
  1.70%   │        0x00007f1607a05759: mov    %r11,0x10(%rbp,%r10,8)
  3.32%   │        0x00007f1607a0575e: mov    %r13,%r11
  0.28%   │        0x00007f1607a05761: imul   0x18(%rbp,%r10,8),%r11
  2.72%   │        0x00007f1607a05767: mov    %r11,0x18(%rbp,%r10,8)
  2.24%   │        0x00007f1607a0576c: mov    %r13,%r11
  1.67%   │        0x00007f1607a0576f: imul   0x20(%rbp,%r10,8),%r11
  0.26%   │        0x00007f1607a05775: mov    %r11,0x20(%rbp,%r10,8)
  6.17%   │        0x00007f1607a0577a: mov    %r13,%r11
  0.69%   │        0x00007f1607a0577d: imul   0x28(%rbp,%r10,8),%r11
  1.54%   │        0x00007f1607a05783: mov    %r11,0x28(%rbp,%r10,8)  ;*lastore {reexecute=0 rethrow=0 return_oop=0}
          │                                                      ; - jpountz.PForDeltaDecoder::prefixSumOfTwoLoops@30 (line 36)
          │                                                      ; - jpountz.PForDeltaDecoder::decodeAndPrefixSum@32 (line 59)
          │                                                      ; - jpountz.PackedIntsDeltaDecodeBenchmark::pForDeltaDecoder@42 (line 29)
          │                                                      ; - jpountz.generated.PackedIntsDeltaDecodeBenchmark_pForDeltaDecoder_jmhTest::pForDeltaDecoder_thrpt_jmhStub@151 (line 240)
  1.80%   │        0x00007f1607a05788: add    $0x4,%r10d         ;*iinc {reexecute=0 rethrow=0 return_oop=0}
          │                                                      ; - jpountz.PForDeltaDecoder::prefixSumOfTwoLoops@31 (line 35)
          │                                                      ; - jpountz.PForDeltaDecoder::decodeAndPrefixSum@32 (line 59)
          │                                                      ; - jpountz.PackedIntsDeltaDecodeBenchmark::pForDeltaDecoder@42 (line 29)
          │                                                      ; - jpountz.generated.PackedIntsDeltaDecodeBenchmark_pForDeltaDecoder_jmhTest::pForDeltaDecoder_thrpt_jmhStub@151 (line 240)
  1.59%   │        0x00007f1607a0578c: cmp    $0x7d,%r10d
          ╰        0x00007f1607a05790: jl     0x00007f1607a05750  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
```
[6]
First run with a single loop:
```
Benchmark                                             (bitsPerValue)  (exceptionCount)  (sameVal)   Mode  Cnt  Score   Error   Units
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder                    0                 0          1  thrpt    5  8.829 ± 0.045  ops/us
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder:·asm               0                 0          1  thrpt         NaN             ---
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder                    0                 0          2  thrpt    5  7.165 ± 0.217  ops/us
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder:·asm               0                 0          2  thrpt         NaN             ---
```
Second run with two loops:
```
Benchmark                                             (bitsPerValue)  (exceptionCount)  (sameVal)   Mode  Cnt  Score   Error   Units
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder                    0                 0          1  thrpt    5  9.294 ± 0.041  ops/us
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder:·asm               0                 0          1  thrpt         NaN             ---
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder                    0                 0          2  thrpt    5  6.635 ± 0.015  ops/us
PackedIntsDeltaDecodeBenchmark.pForDeltaDecoder:·asm               0                 0          2  thrpt         NaN             ---
```

Reply:
So, as of now, I think we leave the implementation as is and hope that we can do something better with more explicit vectorization support in the future. @jpountz / @rmuir does that seem right to you? If you have any suggestions on other was to try to trick this compiler, I'm happy to try them out. And I know you'll call it out if you see something off in my above analysis, since I'm so new to this :)

Reply:
i dont know enough about the autovectorization to trick it. But in general i wonder if we should replace some instances of `ForUtil.BLOCK_SIZE` with `array.length` etc where possible to make the compiler's job easier wrt bounds checks and loop processing. arrays are being passed in as parameters, these are private static methods, so i don't know how smart it is about this today :) On the other hand, maybe it is a waste of your time and just costs a single inexpensive check up front...

Reply:
Interesting thought @rmuir. I'll tweak this to see what kind of difference it makes, but we can't replace `ForUtil.BLOCK_SIZE` with `array.length` in the production code. The array length is actually one more than `ForUtil.BLOCK_SIZE` (as used in `Lucene90PostingsReader`). (See [L317](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90PostingsReader.java#L317) for example). It populates that 129th value with `NO_MORE_DOCS` (i.e., `MAX_INT`) as a end-of-block marker. This was the source of a very frustrating debugging effort on my part while working on this, since early on I was actually using `array.length` instead :)

Reply:
ok, maybe useful for the future to look at. perhaps we could "hold" ForUtil different from postings reader and avoid this. Or maybe you could try something like `array.length & ~(BLOCK_SIZE - 1)` which is similar to what VectorSpecies.loopBound does when writing manually vectorized code. I found it was quite sensitive to this stuff.

Reply:
@rmuir interesting. Yeah, I'd like to explore this further, but I wonder if it makes sense to do so in a follow-up Jira? For starters, intuitively, this case seems pretty uncommon. It will only kick in when all deltas are the same value, but aren't `1`. "Dense" blocks seem like the common case for using 0 bpv, where all deltas would be `1`, and that case is definitely optimized already (`prefixSumOfOnes`). In fact, `ForDeltaUtil` doesn't even use 0 bpv for any case other than `1` (it doesn't actually store the "same value", but rather infers that it's `1` if bpv == 0). So this is already more efficient than what `ForDeltaUtil` is doing for these cases, in the sense that `ForDeltaUtil` would actually fully encode the deltas and go through the whole dane of decoding them, etc.

@rmuir / @jpountz Any concern with me creating a follow-on issue to further investigate and move forward with this PR in its current state?

Reply:
yes definitely followup or whatever. since you asked for suggestions i was just brainstorming... not necessary to be done here.

Reply:
Thanks @rmuir, I went ahead and created [LUCENE-9918](https://issues.apache.org/jira/browse/LUCENE-9918). I appreciate the additional suggestions! This stuff is super interesting and a bit out of my wheelhouse, so I love having more ideas to experiments with :)